rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,435 @@
+
+We also need a way for future Teleport Client requests to retrieve the correct `crypto.PrivateKey`. For RSA keys, we can continue to store them as PEM encoded keys in (`~/.tsh/keys/proxy/user`). For hardware private keys, we will instead store a fake PEM encoded private key which we can use to identity what device and slot to load the private key from.
+
+```bash
+-----BEGIN YUBIKEY PIV PRIVATE KEY-----
+# base64 encoded
+serial_number=<serial_number>
+slot=<slot>
+-----END YUBIKEY PIV PRIVATE KEY-----
+```
+
+#### Supported clients
+
+`tsh` and Teleport Connect will both support hardware private key login, and `tctl` will be able to use resulting login sessions.
+
+#### Unsupported clients
+
+The WebUI will not be able to support PIV login, since it is brower-based and cannot connect directly to the user's PIV device. If a user with `require_session_mfa: hardware_key` attempts to login on the WebUI, or use an existing login session, it will fail. However, WebUI user registration and password reset logic must still work, regardless of the user's private key policy requirement. After initial registration/reset flow, the user should be directed to a page which notifies them that `tsh` or Teleport Connect must be used.
+
+It may be possible to work around this limitation by introducing a local proxy to connect to the hardware key, or by supporting a hardware key solution which doesn't need a direct connection, but this is out of scope and will not be explored in this PR.
+
+In cases where WebUI access is needed or desired, cluster admins should only apply `require_session_mfa: hardware_key | hardware_key_touch` selectively to roles which warrant more protection. Teleport Connect will also serve as a great UI alternative.
+
+### UX
+
+#### Hardware key login
+
+When possible, hardware key login will not be any different from the normal login flow. However, in some cases, additional user intervention will be required. Below are some examples along with the resulting UX.
+
+Note: Teleport Connect will need custom solutions for these edge cases, such as tsh-initiated callbacks.
+
+##### Initial login fails due to an unmet private key policy
+
+```bash
+> tsh login --user=dev
+Enter password for Teleport user dev:
+Tap any security key
+Initial login failed due to an unmet private key policy, "hardware_key".
+Re-initiating login with YubiKey generated private key...
+Enter password for Teleport user dev:
+Tap any security key
+```
+
+Note: this should only occur when a user's role determines it's private key policy requirement, and the user does not have an existing login session which meets the required policy (expired or active).
```

Comment:
For UX, can we be sure we store the key policy when PIV is required regardless of login success? Otherwise user may trigger the unmet key policy again (and the additional roundtrip).

Reply:
> when PIV is required regardless of login success

What do you mean by this? 

To clarify, we only need to do the second round trip on a user's first login with that key policy. Before they attempt to login, we don't have (secure) way to get the user's private key policy requirement since it's determined by their roles. But we do store the key policy in the user's login session so that their next login remembers their required key policy.

Reply:
I mean suppose the login flow is like this:

```bash
> tsh login --user=dev
Cluster "root" requires a YubiKey generated private key to login, but there
is no YubiKey connected. Please insert a YubiKey to re-initiate login...
// tsh polls the PIV library until the user connects a YubiKey (30 second timeout) or the user cancels

...(login attempt times out or is cancelled)...

> tsh login --user=dev
Enter password for Teleport user dev:
Tap any security key
Initial login failed due to an unmet private key policy, "hardware_key".
Re-initiating login with YubiKey generated private key...
Enter password for Teleport user dev:
Tap any security key
```

Would we be able to save the private key policy so that the second login attempt guesses correctly that a yubikey generated private key is required, even though the first login attempt failed?

Reply:
Ah good point, I'll create a fix for this edge case.